### ISSUE TITLE
Add jOOQ-kotlin extension methods to help ignore nullability when mapping

### ISSUE THREAD

**Issue by strindberg** [jOOQ/jOOQ#12934]
This is a side-track/continuation of the discussion in issue #10212. We're using the functional mapping style introduced in jOOQ 3.16. If columns could be referred to in a way that would indicate that they are indeed non-null, Kotlin code using data classes could be simplified.

Consider the following table:

```sql
CREATE TABLE event
(
    id         bigint NOT NULL GENERATED BY DEFAULT AS IDENTITY,
    message    text NOT NULL,
    description text,
    PRIMARY KEY (id)
);
```

This maps to our data class:
```kotlin
data class Event(
    val id: Long,
    val message: String,
    val description: String?
)
```

This is what our select looks like:
```kotlin
dsl.select(
   EVENT.ID,
   EVENT.MESSAGE,
   EVENT.DESCRIPTION,
).from(EVENT).where(EVENT.ID.eq(eventId)).fetchOne(mapping(::toEvent))

private fun toEvent(id: Long?, message: String?, description: String?) = Event(id!!, message!!, description)
```

If there were some null-safe way of referring to the columns we select, we could get rid of the `toEvent` method and refer directly to the constructor in our select:

```kotlin
dsl.select(
   EVENT.ID,
   EVENT.MESSAGE,
    EVENT.DESCRIPTION
).from(EVENT).where(EVENT.ID.eq(eventId)).fetchOne(mapping(::Event))
```

Also note the handling of the `description` column/attribute. It is nullable, and is thus not coerced (with !!) to be non-null as the other attributes. This would have to work as well for this feature to be useful.

**Comment by lukaseder** [comment]
Thanks a lot for your feature request. The `KotlinGenerator` and its defensive approach to nullability is a gift that keeps on giving, hm? :) This compilation wouldn't happen with the `JavaGenerator` (where your field expressions would be of type `Field<T!>`, not `Field<T?>`.

Likewise, if we offered the following set of utilities in `jOOQ-kotlin`:

```kotlin
public inline fun <T1, R> mapping(f: Function1<T1, R>): RecordMapper<Record1<T1>, R> = Records.mapping(f)
public inline fun <T1, T2, R> mapping(f: Function2<T1, T2, R>): RecordMapper<Record2<T1, T2>, R> = Records.mapping(f)
public inline fun <T1, T2, T3, R> mapping(f: Function3<T1, T2, T3, R>): RecordMapper<Record3<T1, T2, T3>, R> = Records.mapping(f)
...
```

... you could profit from the same interoperability trick, where you can again work with `T!` types. Would that be of use to you?

**Comment by lukaseder** [comment]
Speaking of extension functions, here's another useful utility that I think some of you might fancy: https://github.com/jOOQ/jOOQ/issues/12945

**Comment by strindberg** [comment]
> ... you could profit from the same interoperability trick, where you can again work with `T!` types. Would that be of use to you?

Yes, I think this would work very well for us. I realize that the ultimate goal - that jOOQ somehow figures out nullability resulting from every possible query - is unattainable. This proposed solution would put the responsibility for testing our queries for null on us developers, which is fine, but not make our code more cumbersome. So this would definitely make jOOQ more useable for us.

The only thing I would be a little bit wary of is that it _feels_ a little bit unclean: using this "hole" in the java to kotlin interoperability to our advantage. There are now and then weird corner cases when doing this type of smart use of a feature. But I have no concrete example to back that up with.

**Comment by lukaseder** [comment]
So, any feedback on the suggestion of alternative `mapping()` methods, @strindberg, and the upvoters (@sgrimm, @saibotma, @nikolaymetchev, @niamhasura)?

**Comment by lukaseder** [comment]
Oh, sorry, @strindberg, your feedback hadn't loaded in my browser yet...

> The only thing I would be a little bit wary of is that it _feels_ a little bit unclean: using this "hole" in the java to kotlin interoperability to our advantage. There are now and then weird corner cases when doing this type of smart use of a feature. But I have no concrete example to back that up with.

I feel the same way. I mean, if someone *prefers* that (and you all do), then why not just stick with the `JavaGenerator` instead. What does the `KotlinGenerator` buy you?

**Comment by strindberg** [comment]
I will experiment some more with the Java generator. I haven't used it much since it seemed natural to use the Kotlin generator in a Kotlin project.

**Comment by sgrimm** [comment]
On the "why not just use the Java generator" question, the big thing that pushed me toward the Kotlin generator was being able to use named arguments in constructors.

Also, for me the ideal outcome isn't, "Stop the compiler from caring whether or not something can be null," but rather, "Get the benefits of Kotlin's null safety in places where I know the actual value can never be null."

I upvoted the mapping method suggestion not because of the nullability aspect of it, but rather because my code currently uses `convertFrom` in a bunch of places and it'd be nice to have a more concise syntax for that.

**Comment by lukaseder** [comment]
> being able to use named arguments in constructors.

Good point!

> where I know the actual value can never be null.

The really hard part here is to distinguish between *where I know* and *where the compiler can reasonably know*. For the avoidance of doubt, I've laid it out many times, including:

- https://github.com/jOOQ/jOOQ/issues/12949#issuecomment-1023272786
- https://github.com/Kotlin/KEEP/issues/82#issuecomment-1023972766

I still think it's unreasonable to make false promises that cause problems "too often", where the "too often" bit is obviously debatable, and "power users" might be willing to accept more compromises (including starting writing "left join is a code smell" articles 😅).

In any case, the compromise feature request https://github.com/jOOQ/jOOQ/issues/10212 has a high priority for me, so at least, users will have a choice of how to shoot themselves in the foot in this area, once this is implemented.

> I upvoted the mapping method suggestion not because of the nullability aspect of it, but rather because my code currently uses `convertFrom` in a bunch of places and it'd be nice to have a more concise syntax for that.

So, does this work well for you? https://github.com/jOOQ/jOOQ/issues/12945 You can access it already from github: https://github.com/jOOQ/jOOQ/commit/736ce26f915389e0dd8d72fd54e0270baa360ce7

**Comment by strindberg** [comment]
Just to report back, we have decided to use the Java generator for now, as the simplification of our Kotlin code is worth the cost. Like everyone else, we eagerly anticipate some resolution along the lines of #10212.

**Comment by sgrimm** [comment]
I'd commented that one reason to not switch to the Java generator was to get named constructor parameters. The `copy` method is also a key benefit of Kotlin data classes.

There's a hack to get those things back. For named constructor parameters, you can define a top-level Kotlin wrapper function around the constructor. It has to live in a separate package to disambiguate it from the constructor, but it'll look like a constructor call in the application code. And you can define a `copy` extension method that defines default values for all the parameters.

```kotlin
package com.example.wrappers

import com.example.tables.pojos.MyTable

fun MyTable(id: Int? = null, description: String? = null) = MyTable(id, description)

fun MyTable.copy(id: Int? = this.id, description: String? = this.description) = MyTable(id, description)
```

```kotlin
package com.example

import com.example.tables.pojos.MyTable
import com.example.wrappers.copy
import com.example.wrappers.MyTable

fun example(): MyTable {
  // "Constructor" call with named parameters (really calls the wrapper)
  val original = MyTable(id = 123)
  // POJO with copy extension method
  return original.copy(description = "demo")
}
```

You could choose to make some of those parameters non-nullable; the above example mimics the behavior of the existing Kotlin generator. (But if you're using the Java generator, they return Java objects with platform nullability, rather than Kotlin objects with definitely-nullable properties.) And of course you could choose a different name the constructor wrapper function if you don't care about it feeling like a constructor in the application code.

Not claiming this is a _good_ solution, but it might be a useful middle ground for someone.

**Comment by lukaseder** [comment]
Looking into this now, as both this issue here, and the more generic parent issue #10212 have become tremendously popular.

This issue might be easier to address, as it doesn't generate as many edge cases as #10212. What @strindberg describes is that you'd probably be happy with a simple utility like this:

```kotlin
fun <T1, T2, R> mappingIgnoreNulls(f: Function2<T1, T2, R>): RecordMapper<Record2<T1?, T2?>, R> {
    return mapping { t1, t2 -> f.apply(t1!!, t2!!) }
}

data class R (val i: Int, val s: Date?)
val x: List<R> = ctx.select(RECORD2.C1, RECORD2.C2).from(RECORD2).fetch(mappingIgnoreNulls(::R))
```

This example only shows how it would be done for `Record2`, but this obviously scales to all `Record[N]` and `Function[N]` types for degrees 1 - 22.

In short, there's an alternative version of `Records.mapping()`, that would allow for ignoring nullable columns, turning the `R` constructor of the form `(Int, Date?) -> R` into a function of the form `(Int?, Date?) -> R`, producing a `NullPointerException` in case you do something wrong in the query.

That, along with a flag in the code generator that would produce non-nullable attributes in data classes only (see #10212), would probably ease 70% of the pain people are experiencing in this area?

Apart from the inevitable naming bikeshed (`mappingIgnoreNulls` is a long word), am I overlooking something?

**Comment by lukaseder** [comment]
To make things more composable, an additional level of indirection is added, like this:

```kotlin
fun <T1, T2, E> ignoreNulls(f: Function2<T1, T2, E>): Function2<T1?, T2?, E> = 
    t1, t2 -> f.apply(t1!!, t2!!)
```

And then, for improved discoverability:

```kotlin
fun <T1, T2, E> mappingIgnoreNulls(f: Function2<T1, T2, E>): RecordMapper<Record2<T1?, T2?>, E> = 
    Records.mapping(ignoreNulls(f))

@JvmName("mappingRecordIgnoreNulls")
fun <T1, T2, E> Field<Record2<T1?, T2?>>.mappingIgnoreNulls(f: Function2<T1, T2, E>): Field<E> = 
    mapping(ignoreNulls(f))

fun <T1, T2, E> Field<Result<Record2<T1?, T2?>>>.mappingIgnoreNulls(f: Function2<T1, T2, E>): Field<List<E>> = 
    mapping(ignoreNulls(f))

fun <T1, T2, E> Row2<T1?, T2?>.mappingIgnoreNulls(f: Function2<T1, T2, E>): SelectField<E> = mapping(ignoreNulls(f))
```

I like this approach to solving this particular issue here. It does not stand in contrast with #10212, when (possibly) a future jOOQ will add the non-null guarantee as an option on `TableField<R, T>` types, to avoid declaring them as `TableField<R, T?>` as today. In such a case, usres can continue using the `ignoreNulls()` adapter, or just remove it. As such, the approach will be forward compatible.

**Comment by lukaseder** [comment]
I've submitted a draft, that might be published in jOOQ 3.18:
https://github.com/jOOQ/jOOQ/commit/1629cb275120e954ba69700c4b55a663128d6e66

Early feedback very welcome

**Comment by saibotma** [comment]
Actually a great and simple idea, however what about mapping to POJOs where some properties can be null and others not. In those cases, `mappingIgnoreNulls` would throw a null pointer, wouldn't it?

**Comment by lukaseder** [comment]
> In those cases, `mappingIgnoreNulls` would throw a null pointer, wouldn't it?

Yes. That's what you folks want, no? You want to ignore nullability at the interface of SQL and kotlin.

**Comment by strindberg** [comment]
I don't know if we're misunderstanding each other here, but in my issue description, I defined the Event data class like this:

```kotlin
data class Event(
    val id: Long,
    val message: String,
    val description: String?
)
```

In this case, I would not want the mapping to fail if the `description` column was null in the database, since it is marked as nullable.

**Comment by lukaseder** [comment]
> In this case, I would not want the mapping to fail if the `description` column was null in the database, since it is marked as nullable.

Why would it fail?

**Comment by strindberg** [comment]
I could very well be misunderstanding something here, but wouldn't this function be the one called in the current proposal: 

```kotlin
fun <T1, T2, T3, E> ignoreNulls(f: Function3<T1, T2, T3, E>): Function3<T1?, T2?, T3?, E> = Function3<T1?, T2?, T3?, E> { t1: T1?, t2: T2?, t3: T3? -> f.apply(t1!!, t2!!, t3!!) }
```

And would that not mean that the `f.apply(t1!!, t2!!, t3!!)` part would try to coerce the third argument (which could be null) to be not-null, which would make it fail?

**Comment by lukaseder** [comment]
You're right, my bad. I thought the type information was erased completely. But there does seem to be a null check being generated in the lambda...

Well then, back to where we were. 😅 (you know you folks could also try upvoting this feature request here: https://github.com/Kotlin/KEEP/issues/82)

**Comment by lukaseder** [comment]
Obviously I could write that `ignoreNulls()` function in Java, isntead of in kotlin, to prevent the NPE...

**Comment by lukaseder** [comment]
I mean this:

```java
    static <T1, E> Function1<@Nullable T1, E> ignoreNulls(Function1<T1, E> f) {
        return f;
    }

    static <T1, T2, E> Function2<@Nullable T1, @Nullable T2, E> ignoreNulls(Function2<T1, T2, E> f) {
        return f;
    }

    static <T1, T2, T3, E> Function3<@Nullable T1, @Nullable T2, @Nullable T3, E> ignoreNulls(Function3<T1, T2, T3, E> f) {
        return f;
    }
```

Thoughts?

**Comment by morki** [comment]
Maybe you can just write it in Kotlin as:

```kotlin
@Suppress("UNCHECKED_CAST")
fun <T1, E> ignoreNulls(f: Function1<T1, E>): Function1<T1?, E> = Function1<T1?, E> { t1: T1? -> f.apply(t1 as T1) }
```

Which is exactly what it is doing (casting explicitly nullable T1? to potentionally nullable T1).

**Comment by lukaseder** [comment]
Thanks for the hint, @morki. I'll try that soon. This does seem to work:

```kotlin
fun <T> cast(t: T?): T = t as T
fun main() {
    val s: String? = null
    println(cast(s))
}
```

However, this produces a `NPE`:

```kotlin
inline fun <reified T> cast(t: T?): T = t as T
fun main() {
    val s: String? = null
    println(cast(s))
}
```

My experience with kotlin is insufficient to judge whether the former works by design or if it is an "accident" that might, at some point, assume the latter behaviour.

I kinda feel that reverting to the platform type `T!` might be more sound for now, as it represents a truly "unspecified nullability," the desired third nullability type state that jOOQ should usually work with.

**Comment by morki** [comment]
Yeah, that's fine, it would be awesome whatever the implementation will be. :)

**Comment by IvanPavlov1995** [comment]
Faced the same problem during my jooq research for potential use in my team. Made a small workaround that fits my use cases:
```kotlin
@Suppress("UNCHECKED_CAST")
fun <R : Record, T> TableField<R, T?>.asNonNullField() = this as TableField<R, T>
```
Call site: 
```kotlin
data class EntityDto(val id: UUID)

create
  .select(ENTITIES.ID.asNonNullField())
  .from(ENTITIES)
  .fetch(mapping(::EntityDto))
```

So, I got a question, why TableFields are generated with nullable second generic type for columns declared with not null constraint?
